debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java

```diff
@@ -0,0 +1,333 @@
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.transforms.Transformation;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static org.apache.kafka.connect.transforms.util.Requirements.requireStruct;
+
+/**
+ * A logical table consists of one or more physical tables with the same schema. A common use case is sharding -- the
+ * two physical tables `db_shard1.my_table` and `db_shard2.my_table` together form one logical table.
+ *
+ * This Transformation allows us to change a record's topic name and send change events from both physical tables to
+ * one topic. For instance, we might choose to send the two tables from the above example to the topic
+ * `db_shard.my_table`. The config options {@link #LOGICAL_TABLE_REGEX} and {@link #LOGICAL_TABLE_REPLACEMENT} are used
+ * to change the record's topic.
+ *
+ * Now that multiple physical tables can share a topic, the Key Schema can no longer consist of solely the
+ * record's primary / unique key fields, since they are not guaranteed to be unique across tables. We need
+ * some identifier added to the key that distinguishes the different physical tables. The field name specified by
+ * the config option {@link #PHYSICAL_TABLE_IDENTIFIER_FIELD_NAME} (or {@link #DEFAULT_PHYSICAL_TABLE_IDENTIFIER_FIELD_NAME}
+ * if not specified) is added to the key schema for this purpose. By default, its value will be the old topic name, but
+ * if a custom value is desired, the config options {@link #PHYSICAL_TABLE_REGEX} and {@link #PHYSICAL_TABLE_REPLACEMENT}
+ * are used to change it. For instance, in our above example, we might choose to make the identifier `db_shard1` and
+ * `db_shard2` respectively.
+ *
+ * @author David Leibovic
+ */
+public class ByLogicalTableRouter<R extends ConnectRecord<R>> implements Transformation<R> {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private Pattern logicalTableRegex;
+    private String logicalTableReplacement;
+    private Pattern physicalTableRegex;
+    private String physicalTableReplacement;
+    private String physicalTableIdentifierFieldName;
+    private Cache<Schema, Schema> keySchemaUpdateCache;
+    private Cache<Schema, Schema> envelopeSchemaUpdateCache;
+    private final AvroValidator schemaNameValidator = AvroValidator.create(logger);
+
+    private static final Field LOGICAL_TABLE_REGEX = Field.create("logical.table.regex")
+            .withDisplayName("Logical table regex")
+            .withType(ConfigDef.Type.STRING)
+            .withWidth(ConfigDef.Width.LONG)
+            .withImportance(ConfigDef.Importance.LOW)
+            .withValidation(Field::isRequired, Field::isRegex)
+            .withDescription("The regex used for matching the name of the logical table. This regex will be matched " +
+                "against the original topic name.");
```

Comment:
How about simplifying the description a bit more:

    The regex used for extracting the name of the logical table from the original topic name.